### PR TITLE
Pre-release based on tests

### LINKED ISSUE #33: Make pre-release only after tests succeed
Test and pre-release workflow should be combined so that pre-release is only made if the tests are a success.

Also, in my opinion dev pre-release should only be made when master/main branch is updated, perhaps a different tag for the latest pull request?

### PR DESCRIPTION
**Pull request by Rikuoja**
Fix #33 

Will create one release on the master branch head, and another one for the latest PR, but only if the tests pass.

image

### CODE REVIEW

File: infrastructure/template/root/.github/workflows/test-and-pre-release.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
     strategy:
       matrix:
         # Remove unsupported versions and add more versions. Use LTR version in the cov_tests job
-        docker_tags: [release-3_10, latest]
+        docker_tags: [release-3_16, latest]
```
Comment (Joonalai) [inline infrastructure/template/root/.github/workflows/test-and-pre-release.yml:19]: 3.10 is still marked as a minimum version in metadata.txt. Basically the API is still compatible with QGIS 3.10, so maybe we don't want to drop it off the template just yet. (In the future, sure!)

File: infrastructure/template/root/.github/workflows/test-and-pre-release.yml [hunk 2]
```diff
@@ -77,3 +77,53 @@ jobs:
...
           python-qgis-ltr.bat -m pip install -q pytest
           python-qgis-ltr.bat -m pytest -v
+
+  pre-release:
+    name: "Pre Release"
+    runs-on: "ubuntu-latest"
+    needs: [linux_tests, windows_tests]
+
+    steps:
+      - uses: hmarr/debug-action@v2
```
Comment (Joonalai) [inline infrastructure/template/root/.github/workflows/test-and-pre-release.yml:87]: This should probably be removed in the production.

**Review by Joonalai — CHANGES_REQUESTED** [review]
Looks good! Just a few minor things. 

Here are few general ones:
- Maybe consider adding the coverage step as conditional for latest tag in Linux tests to avoid repetition
- Refactor also the badges in the README.md of the template

File: infrastructure/template/root/.github/workflows/test-and-pre-release.yml (hunk 2 above)
Comment (Rikuoja) [reply infrastructure/template/root/.github/workflows/test-and-pre-release.yml:87]: Well, depends on how much time you want to use debugging it every time you want to change it :D

File: infrastructure/template/root/.github/workflows/test-and-pre-release.yml (hunk 1 above)
Comment (Rikuoja) [reply infrastructure/template/root/.github/workflows/test-and-pre-release.yml:19]: Ah okay, will revert this!

File: infrastructure/template/root/.github/workflows/test-and-pre-release.yml (hunk 2 above)
Comment (Rikuoja) [reply infrastructure/template/root/.github/workflows/test-and-pre-release.yml:87]: In general, the lack of verbosity IMO is a problem for GitHub Actions in general, this one at least provides some basic output on every run.

**Comment by Rikuoja** [comment]
Combined codecov with other tests.

In other news, codecov doesn't work anyway (
```{'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}```
), but it was already broken when I entered the room :P

**Comment by Rikuoja** [comment]
> * Refactor also the badges in the README.md of the template

What exactly should be done with the badges, there only seems to be one for actual release, not pre-release?

**Comment by Joonalai** [comment]
> > * Refactor also the badges in the README.md of the template
> 
> What exactly should be done with the badges, there only seems to be one for actual release, not pre-release?

I meant chancing the correct url to the badge mainly :D

**Comment by Joonalai** [comment]
> Combined codecov with other tests.
> 
> In other news, codecov doesn't work anyway (
> `{'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}`
> ), but it was already broken when I entered the room :P

This happens because the repo you are using is not public. Of course it would be best to skip the coverage if the repo is private.

**Comment by Rikuoja** [comment]
> I meant chancing the correct url to the badge mainly :D

Ah but the name of the workflow is still Tests, so the badge should work, or should we rename the workflow? 🤔

**Comment by Joonalai** [comment]
> > I meant chancing the correct url to the badge mainly :D
> 
> Ah but the name of the workflow is still Tests, so the badge should work, or should we rename the workflow? thinking

True dat. Let's keep the old name, so there is nothing to do.

**Comment by Rikuoja** [comment]
> This happens because the repo you are using is not public. Of course it would be best to skip the coverage if the repo is private and add a comment to the workflow expaining that.

Added skipping coverage in case of private repo!

**Pull request merged**